Make std tests pass on newer Android

pcc opened the pull request:
Newer versions of Android forbid the creation of hardlinks as well as Unix domain sockets in the /data filesystem via SELinux rules, which causes several tests depending on this behavior to fail. So let's skip these tests on Android if we see an EACCES from one of these syscalls. To achieve this, introduce a macro with the horrible name of or_panic_or_skip_on_android_eacces (better suggestions welcome) which skips (returns from) the test if an EACCES return value is seen on Android.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @m-ou-se (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

pcc commented:
cc @chriswailes

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3200765041/jobs/5228042834) [(plain)](https://github.com/rust-lang/rust/commit/dc41a8f2cbb2315a3b73ccfe5ad0b07913c33bdf/checks/8754026561/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
    Finished release [optimized] target(s) in 8.19s
tidy check
tidy: Skipping binary file check, read-only filesystem
Checking which error codes lack tests...
tidy error: /checkout/library/std/src/macros.rs:380: platform-specific cfg: cfg(target_os = "android")
* highest error code: E0790
Found 506 error codes
Found 0 error(s) in error codes
Done!
```

</details>

pcc commented:
Unfortunately Android devices do not have a filesystem mounted at `/tmp`. `/data` is AFAIK pretty much the only filesystem that we can rely on to be present and writable as the shell user (which `adb shell` commands run as).

m-ou-se commented:
An issue with this is that we no longer know which tests are actually run and which are skipped on Android. Maybe we should explicitly check for the android version to decide whether a test should be skipped? Or just #[ignore] the tests on Android entirely?

pcc committed:
Make std tests pass on newer Android

Newer versions of Android forbid the creation of hardlinks as well as
Unix domain sockets in the /data filesystem via SELinux rules, which
causes several tests depending on this behavior to fail. So let's
skip these tests on Android with an #[ignore] directive.

pcc commented:
> An issue with this is that we no longer know which tests are actually run and which are skipped on Android. Maybe we should explicitly check for the android version to decide whether a test should be skipped? Or just #[ignore] the tests on Android entirely?

I'm not sure if it's as simple as checking the Android version (I'm not sure when these restrictions were added and it might not have been at the same time on every device). Applications wouldn't be able to use these APIs successfully so there's less value in testing them, but on the other hand Rust is increasingly being used in the operating system itself. If we don't care that much about whether these tests pass on Android and we expect the existing testing on non-Android Linux to be sufficient then yes, the simplest approach is to ignore them on Android, and I've updated the patch to do that. For the operating system use case we may consider arranging to run these tests only on rooted devices, but that can come later.

pcc commented:
Anything else needed here? As I mentioned, the tests are now ignored on Android.

pcc commented:
Ping.

the8472 commented:
@rustbot ready

bors commented:
:umbrella: The latest upstream changes (presumably #106673) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

pcc commented:
Ping.

ChrisDenton commented:
It seems somewhat unfortunate to be ignoring the tests entirely but...

> Applications wouldn't be able to use these APIs successfully so there's less value in testing them, but on the other hand Rust is increasingly being used in the operating system itself... For the operating system use case we may consider arranging to run these tests only on rooted devices, but that can come later.

Fair enough. And ignored tests can be run explicitly. I wonder though if there could be some way for the tests to be opt-in as a group? I'm thinking something like:

```rust
#[cfg_attr(all(target_os = "android", not(test_rooted_android_device)), ignore)]
```

Or some other `--cfg` that can be passed in using rustflags. I'm not sure of the name.

pcc commented:
What I had in mind for later was that the tests could opt in automatically by checking for `getuid() == 0` and skipping if false. That way, we don't end up adding a developer-facing `--cfg` that we'd just remove again a few months later. Choosing the right thing to do here may take some discussion.

I do think though that as a first step we should make these tests stop failing on Android, since after several months it's already getting rather tedious to need to patch this in every time I need to run the tests.

workingjubilee commented:
Extremely annoying but at the same time seems fine to me.
r? @workingjubilee 
@bors r+

bors commented:
:pushpin: Commit fed6fceb6f7f73e6870e84a2a3389cb71108f197 has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee fed6fceb6f7f73e6870e84a2a3389cb71108f197 -->
<!-- homu: {"type":"Approved","sha":"fed6fceb6f7f73e6870e84a2a3389cb71108f197","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit fed6fceb6f7f73e6870e84a2a3389cb71108f197 with merge a6236fa460811bbd4a08a94db249c344fa9f2220...
<!-- homu: {"type":"BuildStarted","head_sha":"fed6fceb6f7f73e6870e84a2a3389cb71108f197","merge_sha":"a6236fa460811bbd4a08a94db249c344fa9f2220"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5666154086/job/15355546847)
Approved by: workingjubilee
Pushing a6236fa460811bbd4a08a94db249c344fa9f2220 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"workingjubilee","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5666154086/job/15355546847"},"merge_sha":"a6236fa460811bbd4a08a94db249c344fa9f2220"} -->

rust-timer commented:
Finished benchmarking commit (a6236fa460811bbd4a08a94db249c344fa9f2220): [comparison URL](https://perf.rust-lang.org/compare.html?start=cf34adb0dd408d7b2f27a64484df63c36c1cc514&end=a6236fa460811bbd4a08a94db249c344fa9f2220&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=cf34adb0dd408d7b2f27a64484df63c36c1cc514&end=a6236fa460811bbd4a08a94db249c344fa9f2220&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=cf34adb0dd408d7b2f27a64484df63c36c1cc514&end=a6236fa460811bbd4a08a94db249c344fa9f2220&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.6% | [-3.7%, -3.6%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=cf34adb0dd408d7b2f27a64484df63c36c1cc514&end=a6236fa460811bbd4a08a94db249c344fa9f2220&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=cf34adb0dd408d7b2f27a64484df63c36c1cc514&end=a6236fa460811bbd4a08a94db249c344fa9f2220&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 650.76s -> 652.439s (0.26%)
